File: js/lib/selfNumber.js
"use strict";

function genNotSelfNum(n) {
  if (n >= 10) {
    nArr = String(n).split('');
    return Number(n) + nArr.reduce(function (acc, v) {
      return acc + Number(v);
    }, 0);
  } else {
    return n + n;
  }
}

function genNotSelfArr(target) {
  var i = 1;
  var arr = [];

  while (i < target) {
    arr.push(genNotSelfNum(i));
    i++;
  }

  return arr;
}

function checkSelfNum(target) {
  arr = genNotSelfArr(target);

  for (var i = 1; i < target; i++) {
    if (arr.indexOf(i) == -1) console.log(i);
  }
}

checkSelfNum(10000);

File: python/self_number.py
from functools import *
import time

def gen_not_self_num(num):
    if num >=10:
        num_arr=list(str(num))
        s_num=int(num)+reduce(lambda a,b:int(a)+int(b),num_arr)
        return int(s_num)
    else:
        return int(num)+int(num)


def gen_not_self_num_arr(target):
    n=1
    not_self_num_arr=[]
    while(n<target):
        not_self_num_arr.append(gen_not_self_num(n))
        n+=1
    return not_self_num_arr

def gen_self_num(target):
    arr=gen_not_self_num_arr(target)
    for i in range(1,target):
        if i not in arr:
            print(i)

start = time.time()
gen_self_num(10000)
print("time:", time.time()-start)
